File: js_version/src/Neuron.js

'use strict'

//
//
// NEURON

// ctor/dtor

function Neuron (numOutputs, myIndex) {

	this._m_layer_index = myIndex;
	this._m_arr_outputWeights = [];

    for (var i = 0; i < numOutputs; ++i)
        this._m_arr_outputWeights.push({
        	  weight: this._randomWeight()
        	, deltaWeight: 0
        });

	// other attr

    this._m_outputVal = 0.0;
    this._m_gradient = 0.0; // used by the backpropagation
};

// static attr

Neuron.prototype._k_eta = 0.15
Neuron.prototype._k_alpha = 0.5

// static inline method(s)

Neuron.prototype._transferFunction = function(x)
{
    return Math.tanh(x); // tanh - output range [-1.0..1.0]
}

Neuron.prototype._transferFunctionDerivative = function(x)
{
    return (1.0 - x * x); // tanh derivative
}

Neuron.prototype._randomWeight = function()
{
    return Math.random();
}

// private method(s)

Neuron.prototype._sumDOW = function(arr_nextLayer)
{
    var sum = 0.0;

    // Sum our contributions of the errors at the nodes we feed.

    var num_neuron = (arr_nextLayer.length - 1); // exclude bias neuron

    for (var n = 0; n < num_neuron; ++n)
        sum += this._m_arr_outputWeights[n].weight * arr_nextLayer[n]._m_gradient;

    return sum;
}

// public method(s)

Neuron.prototype.feedForward = function(arr_prevLayer)
{
    var sum = 0.0;

    // Sum the previous layer's outputs (which are our inputs)
    // Include the bias node from the previous layer.

    for (var n = 0; n < arr_prevLayer.length; ++n)
        sum += arr_prevLayer[n].getOutputVal() * arr_prevLayer[n]._m_arr_outputWeights[this._m_layer_index].weight;

    this._m_outputVal = this._transferFunction(sum);
}

Neuron.prototype.calcOutputGradients = function(targetVal)
{
    var delta = targetVal - this._m_outputVal;
    this._m_gradient = delta * this._transferFunctionDerivative(this._m_outputVal);
}

Neuron.prototype.calcHiddenGradients = function(arr_nextLayer)
{
    var dow = this._sumDOW(arr_nextLayer);
    this._m_gradient = dow * this._transferFunctionDerivative(this._m_outputVal);
}

Neuron.prototype.updateInputWeights = function(arr_prevLayer)
{
    // The weights to be updated are in the Connection container
    // in the neurons in the preceding layer

    for (var n = 0; n < arr_prevLayer.length; ++n)
    {
        var neuron = arr_prevLayer[n];
        var oldDeltaWeight = neuron._m_arr_outputWeights[this._m_layer_index].deltaWeight;

        var newDeltaWeight =
            // Individual input, magnified by the gradient and train rate:
            this._k_eta
            * neuron.getOutputVal()
            * this._m_gradient
            // Also add momentum = a fraction of the previous delta weight;
            + this._k_alpha
            * oldDeltaWeight
            ;

        neuron._m_arr_outputWeights[this._m_layer_index].deltaWeight = newDeltaWeight;
        neuron._m_arr_outputWeights[this._m_layer_index].weight += newDeltaWeight;
    }
}

// getter/setter

Neuron.prototype.setOutputVal = function(val) { this._m_outputVal = val; }
Neuron.prototype.getOutputVal = function() { return this._m_outputVal; }

Neuron.prototype.getOutputWeight = function() { return this._m_arr_outputWeights; }

// NEURON
//
//


module.exports = Neuron;

File: projects/training-logic/src/machine-learning/Neuron.cpp

#include "Neuron.hpp"


#include <cassert>




namespace ActivationFunctions {
    namespace tanh {
        double activation(double x)
        {
            // tanh - output range [-1.0..1.0]
            return std::tanh(x);
        }

        double derivative(double x)
        {
            // tanh derivative

            // faster, less accurate
            // return 1.0 - x * x;

            return 1.0 - std::tanh(x * x);
        }
    }

    namespace relu {
        double activation(double x)
        {
            // relu - output range [0.0..1.0]
            return std::max(x, 0.0);
        }

        double derivative(double x)
        {
            // relu derivative
            // return x < 0.0 ? 0.0 : 1.0;
            return x < 0.0 ? 0.0 : x;
        }
    }

    namespace leakyRelu {
        double activation(double x)
        {
            // leaky relu
            if (x > 0) {
                return x;
            }
            return x * 0.1;
        }

        double derivative(double x)
        {
            // leaky relu derivative
            return x < 0.1 ? 0.0 : x;
        }
    }
}








namespace {
    double k_learningRate = 0.15;  // overall net learning rate, [0.0..1.0]
    double k_alpha = 0.5; // momentum, multiplier of last deltaWeight, [0.0..1.0]
}

Neuron::Neuron(uint32_t numOutputs, uint32_t myIndex, RandomNumberGenerator& rng)
    : _layerIndex(myIndex)
{
    if (numOutputs == 0) {
        return;
    }

    _outputSynapses.reserve(numOutputs);
    for (uint32_t ii = 0; ii < numOutputs; ++ii) {
        _outputSynapses.emplace_back(rng.getRangedValue(0.0f, 1.0f));
    }
}

// #define D_USE_RELU

void Neuron::feedForward(const t_Layer &prevLayer)
{
    double sum = 0.0;

    // Sum the previous layer's outputs (which are our inputs)
    // Include the bias node from the previous layer.

    for (uint32_t ii = 0; ii < prevLayer.size(); ++ii) {
        const auto& currNeuron = prevLayer[ii];

        sum += currNeuron._outputVal * currNeuron._outputSynapses[_layerIndex].weight;
    }

#ifndef D_USE_RELU
    _outputVal = ActivationFunctions::tanh::activation(sum);
#else
    _outputVal = ActivationFunctions::leakyRelu::activation(sum);
#endif
}

void Neuron::calcOutputGradients(double targetVal)
{
#ifndef D_USE_RELU
    const double delta = targetVal - _outputVal;
    _gradientOutputVal = delta * ActivationFunctions::tanh::derivative(_outputVal);
#else
    _gradientOutputVal = 2.0 * (_outputVal - targetVal);
    // const double delta = _outputVal - targetVal;
    // _gradientOutputVal = delta * ActivationFunctions::leakyRelu::derivative(_outputVal);
#endif
}

void Neuron::calcHiddenGradients(const t_Layer &nextLayer)
{
    const double dow = _sumDOW(nextLayer);
#ifndef D_USE_RELU
    _gradientOutputVal = dow * ActivationFunctions::tanh::derivative(_outputVal);
#else
    _gradientOutputVal = dow * ActivationFunctions::leakyRelu::derivative(_outputVal);
#endif
}

void Neuron::updateInputWeights(t_Layer &prevLayer) const
{
    // The weights to be updated are in the Connection container
    // in the neurons in the preceding layer

    for (Neuron &prevNeuron : prevLayer) {

        auto& prevSynapse = prevNeuron._outputSynapses[_layerIndex];

        // const double oldDeltaWeight = prevSynapse.deltaWeight;

        const double newDeltaWeight =
            // Individual input, magnified by the gradient and train rate:
            k_learningRate * prevNeuron._outputVal * _gradientOutputVal
            // // Also add momentum = a fraction of the previous delta weight;
            // + k_alpha * oldDeltaWeight
            ;

        prevSynapse.deltaWeight = newDeltaWeight;
        prevSynapse.weight += newDeltaWeight;
    }
}

double Neuron::_sumDOW(const t_Layer &nextLayer) const
{
    double sum = 0.0;

    // Sum our contributions of the errors at the nodes we feed.

    const uint32_t totalNeurons = uint32_t(nextLayer.size()) - 1; // exclude bias neuron

    for (uint32_t ii = 0; ii < totalNeurons; ++ii) {
        sum += _outputSynapses[ii].weight * nextLayer[ii]._gradientOutputVal;
    }

    return sum;
}
